File: adv22b.nim
import sets, strutils

func score(p: seq[int]): int =
  var m = p.len
  for i in 0..<p.len:
    result += p[i] * m
    dec m

func game(p1, p2: seq[int]): (int, int) =
  var
    seen = initHashSet[(int,int)]()
    p1 = p1
    p2 = p2
  while true:
    let id = (p1.score, p2.score)
    if id in seen or p2.len == 0:
      return (0, id[0])                # Player 1 wins
    if p1.len == 0:
      return (1, id[1])
    seen.incl id
    var winner: int
    if p1[0] < p1.len and p2[0] < p2.len:
      var tmp: int
      (winner, tmp) = game(p1[1..p1[0]], p2[1..p2[0]])
    elif p1[0] > p2[0]:
      winner = 0
    else:
      winner = 1
    if winner == 0:
      p1.add p1[0]
      p1.add p2[0]
    else:
      p2.add p2[0]
      p2.add p1[0]
    p1.delete 0
    p2.delete 0

var
  player1 = true
  p1, p2: seq[int]
for line in "adv22.txt".lines:
  if line.find("Player") != -1:
    continue
  if line == "":
    player1 = false
    continue
  if player1:
    p1.add line.parseInt
  else:
    p2.add line.parseInt

echo game(p1, p2)[1]

File: adv22b.jl
score(pack) = mapreduce(prod, +, enumerate(pack), init=0)

function game(pack1, pack2)
    seen = Set{Tuple{Int,Int}}() # should be Set{Tuple{Vector{Int},Vector{Int}}} but this also works
    pack1, pack2 = copy(pack1), copy(pack2)
    while true
        id = (score(pack1), score(pack2))
        (id in seen || isempty(pack2)) && return (1, id[1])
        isempty(pack1) && return (2, id[2])
        push!(seen, id)
        card1, card2 = pop!(pack1), pop!(pack2)
        if card1 <= length(pack1) && card2 <= length(pack2)
            (winner, _) = game(pack1[end-card1+1:end], pack2[end-card2+1:end])
        else
            winner = card1 > card2 ? 1 : 2
        end
        if winner == 1
            pushfirst!(pack1, card2, card1)
        else
            pushfirst!(pack2, card1, card2)
        end
    end
end

function adv22(filename)
    player1 = true
    pack1::Vector{Int}, pack2::Vector{Int} = [], []
    for line in eachline(filename)
        findfirst("Player", line) != nothing && continue
        if line == ""
            player1 = false
            continue
        end
        pushfirst!(player1 ? pack1 : pack2, parse(Int, line))
    end
    game(pack1, pack2)[2]
end
